Demographic Data Analyzer

In this challenge you must analyze demographic data using Pandas. You are given a dataset of demographic data that was extracted from the 1994 Census database.

You must use Pandas to answer the following questions:

•	How many people of each race are represented in this dataset? This should be a Pandas series with race names as the index labels. (race column)

•	What is the average age of men?

•	What is the percentage of people who have a Bachelor's degree?

•	What percentage of people with advanced education (Bachelors, Masters, or Doctorate) make more than 50K?

•	What percentage of people without advanced education make more than 50K?

•	What is the minimum number of hours a person works per week?

•	What percentage of the people who work the minimum number of hours per week have a salary of more than 50K?

•	What country has the highest percentage of people that earn >50K and what is that percentage?

•	Identify the most popular occupation for those who earn >50K in India.






#Reading database

In [8]:
import pandas as pd

# Load the data with proper formatting
df = pd.read_csv('/content/drive/MyDrive/adult.data.csv', header=None, names=[
    "age", "workclass", "fnlwgt", "education", "education-num",
    "marital-status", "occupation", "relationship", "race", "sex",
    "capital-gain", "capital-loss", "hours-per-week", "native-country", "salary"
], skipinitialspace=True)

# Let's check the data types
print(df.dtypes)
print(df.head())


age               object
workclass         object
fnlwgt            object
education         object
education-num     object
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain      object
capital-loss      object
hours-per-week    object
native-country    object
salary            object
dtype: object
   age         workclass  fnlwgt  education  education-num  \
0  age         workclass  fnlwgt  education  education-num   
1   39         State-gov   77516  Bachelors             13   
2   50  Self-emp-not-inc   83311  Bachelors             13   
3   38           Private  215646    HS-grad              9   
4   53           Private  234721       11th              7   

       marital-status         occupation   relationship   race   sex  \
0      marital-status         occupation   relationship   race   sex   
1       Never-married       Adm-clerical  Not-in-family  White  Male   
2  Married-civ-spouse 

# Race Count

In [9]:
# Count how many people of each race are represented in this dataset
race_count = df['race'].value_counts()

# Display the result
print("Number of each race:\n", race_count)


Number of each race:
 race
White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
race                      1
Name: count, dtype: int64


# Average age of men

In [11]:
df['age'] = pd.to_numeric(df['age'], errors='coerce')
print(df['age'].isna().sum(), "non-numeric entries found in 'age' column.")


1 non-numeric entries found in 'age' column.


In [12]:
average_age_men = df[df['sex'] == 'Male']['age'].mean().round(1)
print("Average age of men:", average_age_men)


Average age of men: 39.4


# People with bachelor's degree

In [13]:
# Calculate the percentage of people who have a Bachelor's degree
percentage_bachelors = (df['education'] == 'Bachelors').mean() * 100
percentage_bachelors = round(percentage_bachelors, 1)

# Display the result
print(f"Percentage with Bachelor's degrees: {percentage_bachelors}%")


Percentage with Bachelor's degrees: 16.4%


# People with advanced education who earn >50K

In [14]:
# Define advanced education
higher_education = df[df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])]

# Calculate percentage of people with advanced education who earn >50K
higher_education_rich = (higher_education['salary'] == '>50K').mean() * 100
higher_education_rich = round(higher_education_rich, 1)

# Display the result
print(f"Percentage with higher education that earn >50K: {higher_education_rich}%")


Percentage with higher education that earn >50K: 46.5%


# People without advanced education who earn >50K

In [15]:
# Define non-advanced education group
lower_education = df[~df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])]

# Calculate percentage of people without advanced education who earn >50K
lower_education_rich = (lower_education['salary'] == '>50K').mean() * 100
lower_education_rich = round(lower_education_rich, 1)

# Display the result
print(f"Percentage without higher education that earn >50K: {lower_education_rich}%")


Percentage without higher education that earn >50K: 17.4%


# Minimum of hours working per week

In [16]:
# Find the minimum number of hours worked per week
min_work_hours = df['hours-per-week'].min()

# Display the result
print(f"Min work time: {min_work_hours} hours/week")


Min work time: 1 hours/week


# Min work-time with max earn

In [17]:
# Filter people who work the minimum number of hours
min_workers = df[df['hours-per-week'] == min_work_hours]

# Calculate the percentage of those earning >50K
rich_percentage = (min_workers['salary'] == '>50K').mean() * 100
rich_percentage = round(rich_percentage, 1)

# Display the result
print(f"Percentage of rich among those who work fewest hours: {rich_percentage}%")


Percentage of rich among those who work fewest hours: 10.0%


# Country with highest percentage of rich:


In [18]:
# Group by country and calculate the percentage of people earning >50K
country_earnings = df.groupby('native-country')['salary'].apply(lambda x: (x == '>50K').mean() * 100)

# Identify the country with the highest percentage
highest_earning_country = country_earnings.idxmax()
highest_earning_country_percentage = round(country_earnings.max(), 1)

# Display the results
print("Country with highest percentage of rich:", highest_earning_country)
print(f"Highest percentage of rich people in country: {highest_earning_country_percentage}%")


Country with highest percentage of rich: Iran
Highest percentage of rich people in country: 41.9%


# Popular occupation for those who earn >50K in India

In [20]:
# Filter for people from India earning >50K
rich_indians = df[(df['native-country'] == 'India') & (df['salary'] == '>50K')]

# Find the most frequent occupation among them
top_IN_occupation = rich_indians['occupation'].mode()[0]

# Display the result
print("Top occupation in India for those who earn >50K:", top_IN_occupation)


Top occupation in India for those who earn >50K: Prof-specialty


# Demographic Data Analyzer Calculator

In [26]:
import pandas as pd

def calculate_demographic_data(print_data=True):
    # Read data from file
    df = pd.read_csv('/content/drive/MyDrive/adult.data.csv', header=None, names=[
        "age", "workclass", "fnlwgt", "education", "education-num",
        "marital-status", "occupation", "relationship", "race", "sex",
        "capital-gain", "capital-loss", "hours-per-week", "native-country", "salary"
    ])

    # Clean whitespace from object (string) columns
    df = df.apply(lambda col: col.str.strip() if col.dtype == 'object' else col)

    # Force numeric columns to correct dtype, coercing errors
    for col in ["age", "education-num", "capital-gain", "capital-loss", "hours-per-week"]:
        df[col] = pd.to_numeric(df[col], errors='coerce')

    # Drop rows with NaNs in important columns
    df.dropna(subset=["age", "hours-per-week", "salary"], inplace=True)

    # How many of each race are represented in this dataset?
    race_count = df['race'].value_counts()

    # What is the average age of men?
    average_age_men = round(df[df['sex'] == 'Male']['age'].mean(), 1)

    # Percentage of people with a Bachelor's degree
    percentage_bachelors = round((df['education'] == 'Bachelors').mean() * 100, 1)

    # Advanced education
    higher_education = df[df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])]
    lower_education = df[~df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])]

    higher_education_rich = round(
        (higher_education['salary'] == '>50K').mean() * 100, 1
    )

    lower_education_rich = round(
        (lower_education['salary'] == '>50K').mean() * 100, 1
    )

    # Minimum hours per week
    min_work_hours = df['hours-per-week'].min()

    # Percentage of rich among people who work minimum hours
    min_workers = df[df['hours-per-week'] == min_work_hours]
    rich_percentage = round(
        (min_workers['salary'] == '>50K').mean() * 100, 1
    )

    # Country with highest % of rich
    country_counts = df['native-country'].value_counts()
    rich_by_country = df[df['salary'] == '>50K']['native-country'].value_counts()
    rich_percentage_country = (rich_by_country / country_counts * 100).dropna()
    highest_earning_country = rich_percentage_country.idxmax()
    highest_earning_country_percentage = round(rich_percentage_country.max(), 1)

    # Top occupation in India for >50K
    top_IN_occupation = df[
        (df['native-country'] == 'India') & (df['salary'] == '>50K')
    ]['occupation'].mode()[0]

    # DO NOT MODIFY BELOW THIS LINE
    if print_data:
        print("Number of each race:\n", race_count)
        print("Average age of men:", average_age_men)
        print(f"Percentage with Bachelors degrees: {percentage_bachelors}%")
        print(f"Percentage with higher education that earn >50K: {higher_education_rich}%")
        print(f"Percentage without higher education that earn >50K: {lower_education_rich}%")
        print(f"Min work time: {min_work_hours} hours/week")
        print(f"Percentage of rich among those who work fewest hours: {rich_percentage}%")
        print("Country with highest percentage of rich:", highest_earning_country)
        print(f"Highest percentage of rich people in country: {highest_earning_country_percentage}%")
        print("Top occupations in India:", top_IN_occupation)

    return {
        'race_count': race_count,
        'average_age_men': average_age_men,
        'percentage_bachelors': percentage_bachelors,
        'higher_education_rich': higher_education_rich,
        'lower_education_rich': lower_education_rich,
        'min_work_hours': min_work_hours,
        'rich_percentage': rich_percentage,
        'highest_earning_country': highest_earning_country,
        'highest_earning_country_percentage': highest_earning_country_percentage,
        'top_IN_occupation': top_IN_occupation
    }

# Run it
calculate_demographic_data()


Number of each race:
 race
White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: count, dtype: int64
Average age of men: 39.4
Percentage with Bachelors degrees: 16.4%
Percentage with higher education that earn >50K: 46.5%
Percentage without higher education that earn >50K: 17.4%
Min work time: 1.0 hours/week
Percentage of rich among those who work fewest hours: 10.0%
Country with highest percentage of rich: Iran
Highest percentage of rich people in country: 41.9%
Top occupations in India: Prof-specialty


{'race_count': race
 White                 27816
 Black                  3124
 Asian-Pac-Islander     1039
 Amer-Indian-Eskimo      311
 Other                   271
 Name: count, dtype: int64,
 'average_age_men': np.float64(39.4),
 'percentage_bachelors': np.float64(16.4),
 'higher_education_rich': np.float64(46.5),
 'lower_education_rich': np.float64(17.4),
 'min_work_hours': 1.0,
 'rich_percentage': np.float64(10.0),
 'highest_earning_country': 'Iran',
 'highest_earning_country_percentage': 41.9,
 'top_IN_occupation': 'Prof-specialty'}

# Conclusions and Insights

In this project, I cleaned and analyzed a real-world demographic dataset using Pandas, implementing logical filters and summary calculations to extract meaningful insights. I structured the entire analysis within a reusable and testable Python function, debugging and optimizing the code step by step. As a result, I was able to uncover key patterns in age, education, work hours, income, and geographic distribution—successfully applying core Data Science techniques in a practical and interpretable way.

Some insights that we can extract from the Dataset are:

- Demographics:The dataset is heavily skewed toward the White population (over 27,000 individuals), which is important to consider in any generalizations.

- Education & Income:

  - People with higher education (Bachelor's or above) are much more likely to earn >50K (46.5%) compared to those without (17.4%).

  - But higher education doesn't guarantee a high income — more than half of highly educated individuals still earn <=50K.

- Work Hours & Income:

  - Some people only work 1 hour per week, and 10% of them still earn >50K — this likely includes highly compensated consultants or self-employed professionals.

- Global Income Disparities:

  - Iran surprisingly has the highest percentage of high-income individuals relative to its population in the dataset. This may be due to small sample size or specific migrant profiles in the U.S.

  - Top Job in India for High Earners: Prof-specialty, reinforcing the idea that specialized, high-skilled professions are key to economic mobility.